src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1965,9 +1984,11 @@ public Task StopAsync()
 
         private async Task StopAsyncCore()
         {
+            _hubConnection._state.AssertInConnectionLock();
+
             Log.Stopping(_logger);
 
-            _messageBuffer?.Dispose();
+            Cleanup();
```

Comment:
What happens if we receive an ack after disposing the message buffer but before stopping the read loop? Can we rely on this always being called by `HandleConnectionClose`? Is there any situation this wouldn't be called after `CancelPendingRead()`?